terraform/registry/server.go

```diff
@@ -0,0 +1,110 @@
+package registry
+
+import (
+	"context"
+	"net/http"
+	"net/url"
+
+	"github.com/gruntwork-io/go-commons/errors"
+	"github.com/gruntwork-io/terragrunt/pkg/log"
+	"github.com/gruntwork-io/terragrunt/terraform/registry/controllers"
+	"github.com/gruntwork-io/terragrunt/terraform/registry/handlers"
+	"github.com/gruntwork-io/terragrunt/terraform/registry/router"
+	"github.com/gruntwork-io/terragrunt/terraform/registry/services"
+	"golang.org/x/sync/errgroup"
+)
+
+// Server is a private Terraform registry for provider caching.
+type Server struct {
+	http.Handler
+	config *Config
+
+	providerService    *services.ProviderService
+	providerController *controllers.ProviderController
+}
+
+// NewServer returns a new Server instance.
+func NewServer(config *Config, providerService *services.ProviderService) *Server {
+	authorization := &handlers.Authorization{
+		Token: config.Token,
+	}
+
+	reverseProxy := &handlers.ReverseProxy{
+		ServerURL: &url.URL{
+			Scheme: "http",
+			Host:   config.Addr(),
+		},
+	}
+
+	downloaderController := &controllers.DownloaderController{
+		ReverseProxy:    reverseProxy,
+		ProviderService: providerService,
+	}
+
+	providerController := &controllers.ProviderController{
+		Authorization:   authorization,
+		ReverseProxy:    reverseProxy,
+		ProviderService: providerService,
+		Downloader:      downloaderController,
+	}
+
+	discoveryController := &controllers.DiscoveryController{
+		Endpointers: []controllers.Endpointer{providerController},
+	}
+
+	rootRouter := router.New()
+	rootRouter.Register(discoveryController, downloaderController)
+
+	v1Group := rootRouter.Group("v1")
+	v1Group.Register(providerController)
+
+	return &Server{
+		Handler:            rootRouter,
+		config:             config,
+		providerService:    providerService,
+		providerController: providerController,
+	}
+}
+
+// ProviderURL returns a full URL to the provider controller, e.g. http://localhost:5758/v1/providers
+func (server *Server) ProviderURL() *url.URL {
+	return &url.URL{
+		Scheme: "http",
+		Host:   server.config.Addr(),
+		Path:   server.providerController.Path(),
+	}
+}
+
+// Run starts the webserver and workers.
+func (server *Server) Run(ctx context.Context) error {
```

Comment:
What happens if you run multiple TG commands concurrently on the same system? Will they clash because they both try to fire up a server on the same port?

Reply:
It was like that. Now the free port of the port is resolved automatically.